shopfloor_single_product_transfer/services/single_product_transfer.py

```diff
@@ -0,0 +1,649 @@
+        search = self._actions_for("search")
+        handlers = [
+            self._scan_product__select_move_line,
+            # If no line is found, we might try to create one,
+            # only if allow_move_create option is True
+            self._scan_product__check_create_move_line,
+            # And unreserve other moves before,
+            # if allow_unreserve_other_moves option is set to True.
+            self._scan_product__unreserve_move_line,
+            self._scan_product__create_move_line,
+        ]
+        lot = search.lot_from_scan(barcode)
+        if lot:
+            product = lot.product_id
+            product_response = self._use_handlers(handlers, product, location, lot=lot)
+            if product_response:
+                return product_response
+
+    def _use_handlers(self, handlers, *args, **kwargs):
+        # TODO: each handler should raise a Shopfloor dedicated exception
+        # with the response data attached
+        for handler in handlers:
+            response = handler(*args, **kwargs)
+            if response:
+                return response
+
+    # Copied from manual_product_transfer
+    def _find_location_move_lines_domain(self, location, product, lot=None):
+        domain = [
+            ("location_id", "=", location.id),
+            ("product_id", "=", product.id),
+            ("state", "in", ("assigned", "partially_available")),
+            ("picking_id.user_id", "in", (False, self.env.uid)),
+        ]
+        if lot:
+            domain = AND([domain, [("lot_id", "=", lot.id)]])
+        return domain
+
+    # Copied from manual_product_transfer
+    def _find_location_move_lines(self, location, product, lot=None):
+        """Find existing move lines in progress related to the source location
+        but not linked to any user.
+        """
+        domain = self._find_location_move_lines_domain(location, product, lot=lot)
+        return self.env["stock.move.line"].search(domain)
+
+    # Copied from manual_product_transfer
+    def _create_move_from_location(self, location, product, quantity, lot=None):
+        picking_type = self.picking_types
+        move_vals = {
+            "name": product.name,
+            "company_id": picking_type.company_id.id,
+            "product_id": product.id,
+            "product_uom": product.uom_id.id,
+            "product_uom_qty": quantity,
+            "location_id": location.id,
+            "location_dest_id": picking_type.default_location_dest_id.id,
+            "origin": self.work.menu.name,
+            "picking_type_id": picking_type.id,
+        }
+        move = self.env["stock.move"].create(move_vals)
+        move._action_confirm(merge=False)
+        move.with_context(
+            {"force_reservation": self.work.menu.allow_force_reservation}
+        )._action_assign()
+        assert move.state == "assigned", "The reservation of quantities has failed"
+        move_line = move.move_line_ids
```

Comment:
This won't work if the move reserved for instance 3 packages + a bunch of products w/o lot/package. In this case you'll got 4 move lines (1 line per package + 1 line for bulk products).

@jbaudoux should we check earlier the quantity we computed in the location for the given product/lot(/package)? I mean:

- if we scan a product and we have 1 PKG + some bulk products in the location, we move only the bulk products (so we can edit the qty to move on our created move line, and make use of the `no_prefill_qty` option)
- if we scan a package (to implement later), we move only this package

What we want to do is to get only one move line for the created move, so we can edit the qty to move. In the case of moving a package, we should even prevent to edit the qty to move so we move it entirely.

Reply:
I'm not sure to understand. What is to be done here?

Reply:
Right now, at least add a comment above this line saying we expect to get only one move line as we are moving only bulk products w/o lot or package.

Reply:
This would be safer then
`move_line = move.move_line_ids[0]`